File: src/chunker.ts
import type { MsgRow } from "./store.js";

// 参数依据实测数据校准（手册附录 B）：
// 私聊消息平均仅 6 字符、中位间隔 6 秒、全天 >30min 间隔仅 6 次
// → 字符/消息数上限才是实际的主切分器，30min GAP 只负责天然会话边界
const GAP_MINUTES = 30;
const MAX_CHARS = 800;
const MAX_MSGS = 60;
const MIN_CHARS = 50;
const MERGE_GAP_HOURS = 2;
const OVERLAP_MSGS = 3;

export interface Chunk {
  thread: string;
  start_time: string;
  end_time: string;
  participants: string; // JSON 数组
  msg_ids: string;      // JSON 数组
  text: string;
  summary: string | null;
}

const ts = (m: MsgRow) => new Date(m.timestamp + "Z").getTime(); // 仅作差值用，时区偏移可抵消
const chars = (msgs: MsgRow[]) => msgs.reduce((n, m) => n + [...m.content].length, 0);

export function chunkThread(thread: string, msgs: MsgRow[], threadType = ""): Chunk[] {
  if (msgs.length === 0) return [];

  // 1) 顺序扫描切块
  const raw: MsgRow[][] = [];
  let cur: MsgRow[] = [];
  let curChars = 0;
  for (const m of msgs) {
    const len = [...m.content].length;
    const gapMin = cur.length > 0 ? (ts(m) - ts(cur[cur.length - 1])) / 60000 : 0;
    if (cur.length > 0 && (gapMin > GAP_MINUTES || curChars + len > MAX_CHARS || cur.length >= MAX_MSGS)) {
      raw.push(cur);
      cur = [];
      curChars = 0;
    }
    cur.push(m);
    curChars += len;
  }
  if (cur.length > 0) raw.push(cur);

  // 2) 小块合并：字符数 < MIN_CHARS 且与下一块间隔 < 2h → 并入下一块
  //    （避免"好的""收到"单独成块）
  const merged: MsgRow[][] = [];
  for (let i = 0; i < raw.length; i++) {
    const block = raw[i];
    const next = raw[i + 1];
    if (
      next &&
      chars(block) < MIN_CHARS &&
      (ts(next[0]) - ts(block[block.length - 1])) / 3600000 < MERGE_GAP_HOURS
    ) {
      raw[i + 1] = [...block, ...next];
      continue;
    }
    merged.push(block);
  }

  // 3) 滑动重叠：每块开头复制上一块末尾 3 条（仅进入文本，不进入 msg_ids），
  //    缓解话题恰好在块边界切断的情况
  return merged.map((block, i) => {
    const overlap = i > 0 ? merged[i - 1].slice(-OVERLAP_MSGS) : [];
    const participants = [...new Set(block.map((m) => m.sender))];
    const start = block[0].timestamp;
    const end = block[block.length - 1].timestamp;
    const header = `[${start.replace("T", " ").slice(0, 16)} ~ ${end.replace("T", " ").slice(11, 16)}] ${threadType}${thread}（${participants.join("、")}）`;
    const lines = [...overlap, ...block].map((m) => `${m.sender}: ${m.content}`);
    return {
      thread,
      start_time: start,
      end_time: end,
      participants: JSON.stringify(participants),
      msg_ids: JSON.stringify(block.map((m) => m.id)),
      text: `${header}\n${lines.join("\n")}`,
      summary: null,
    };
  });
}

File: core/chunker.py
from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import datetime
from typing import Any


GAP_MINUTES = 30
MAX_CHARS = 800
MAX_MESSAGE_CHARS_IN_CHUNK = 1200
MAX_MSGS = 60
MIN_CHARS = 50
MERGE_GAP_HOURS = 2
OVERLAP_MSGS = 3


@dataclass(frozen=True)
class Chunk:
    thread: str
    start_time: str
    end_time: str
    participants: str
    msg_ids: str
    text: str
    summary: str | None = None


def _ts(message: Any) -> float:
    return datetime.fromisoformat(message["timestamp"]).timestamp()


def _chars(messages: list[dict[str, Any]]) -> int:
    return sum(len(message["content"]) for message in messages)


def _chunk_content(content: Any) -> str:
    text = str(content or "")
    if len(text) <= MAX_MESSAGE_CHARS_IN_CHUNK:
        return text
    omitted = len(text) - MAX_MESSAGE_CHARS_IN_CHUNK
    return f"{text[:MAX_MESSAGE_CHARS_IN_CHUNK]}...[已截断 {omitted} 字，使用 get_context 查看原文]"


def _format_chunk_time_range(start: str, end: str) -> str:
    start_label = start.replace("T", " ")[:16]
    end_label = end.replace("T", " ")[:16]
    if start_label[:10] == end_label[:10]:
        end_label = end_label[11:16]
    return f"{start_label} ~ {end_label}"


def _format_message_lines(message: dict[str, Any]) -> str:
    sender = str(message["sender"])
    content = _chunk_content(message["content"])
    lines = content.splitlines() or [""]
    return "\n".join(f"{sender}: {line}" for line in lines)


def chunk_thread(thread: str, messages: list[dict[str, Any]], thread_type: str = "") -> list[Chunk]:
    if not messages:
        return []
    messages = sorted(messages, key=lambda message: (message["timestamp"], str(message.get("id", ""))))

    raw: list[list[dict[str, Any]]] = []
    current: list[dict[str, Any]] = []
    current_chars = 0

    for message in messages:
        length = len(message["content"])
        gap_min = (_ts(message) - _ts(current[-1])) / 60 if current else 0
        should_cut = (
            current
            and (gap_min > GAP_MINUTES or current_chars + length > MAX_CHARS or len(current) >= MAX_MSGS)
        )
        if should_cut:
            raw.append(current)
            current = []
            current_chars = 0
        current.append(message)
        current_chars += length

    if current:
        raw.append(current)

    merged: list[list[dict[str, Any]]] = []
    i = 0
    while i < len(raw):
        block = raw[i]
        next_block = raw[i + 1] if i + 1 < len(raw) else None
        if next_block and _chars(block) < MIN_CHARS and (_ts(next_block[0]) - _ts(block[-1])) / 3600 < MERGE_GAP_HOURS:
            raw[i + 1] = [*block, *next_block]
            i += 1
            continue
        if (
            not next_block
            and merged
            and _chars(block) < MIN_CHARS
            and (_ts(block[0]) - _ts(merged[-1][-1])) / 3600 < MERGE_GAP_HOURS
        ):
            merged[-1] = [*merged[-1], *block]
            i += 1
            continue
        merged.append(block)
        i += 1

    chunks: list[Chunk] = []
    for idx, block in enumerate(merged):
        overlap = merged[idx - 1][-OVERLAP_MSGS:] if idx > 0 else []
        participants = list(dict.fromkeys(message["sender"] for message in block))
        start = block[0]["timestamp"]
        end = block[-1]["timestamp"]
        header = f"[{_format_chunk_time_range(start, end)}] {thread_type}{thread}（{'、'.join(participants)}）"
        lines = [_format_message_lines(message) for message in [*overlap, *block]]
        chunks.append(
            Chunk(
                thread=thread,
                start_time=start,
                end_time=end,
                participants=json.dumps(participants, ensure_ascii=False),
                msg_ids=json.dumps([message["id"] for message in block], ensure_ascii=False),
                text=f"{header}\n" + "\n".join(lines),
            )
        )

    return chunks
